### PR TITLE
Noiseaware CSPLayout pass

### LINKED ISSUE #5052: CSPLayout should be noise-aware
### What is the expected behavior?

When observing the time evolution of simple spin chains, I encountered an issue with the CSPLayout. In case a circuit is mapped to a device that allows for multiple exact (non swap introducing) mappings of this circuit, the CSPLayout method does not always choose the best solution. It chooses the one which occurs first in its solution process. In such a case it should provide the solution with the least noise. Therefore, I suggest to make CSPLayout noise aware and will try to work on it.

### PR DESCRIPTION
**Pull request by czachow**
### Summary
Refactored the class CSPLayout and created a new class derived of CSPLayout which adds noise awareness.

Closes #5052

### Details and comments
- refactored the custom recursive backtracking solver into a separate document
- refactored the method run of CSPLayout into different methods
- created a new method called CSPLayoutNoise
- implemented tests for CSPLayoutNoise

### CODE REVIEW

**Comment by czachow** [comment]
Closes #5072

**Comment by 1ucian0** [comment]
It could be nice to have some performance/benchmarking data, if it's not too hard to get.

**Comment by czachow** [comment]
Thanks for your feedback. I took a look into the other layouters and think it is not easy to compare them. Here you find some benchmarking comparison:

As a benchmark I took the problem of simulating the time evolution of an Ising Chain. The calculation was run on ibmq_bogota using three qubits. Three different layouting methods were used and compared to the simulator. At first the classical CSPLayout (dev), the CSPLayoutNoise (dev_ns) as implemented here and a more advanced CSPLayoutAdvancedNoise (dev_avns) as described in slack were used. The plot shows the probability of measuring the state 111, starting with an initial state of 000 after a certain time. Additionally, the average of five "independent" runs have been taken. This shows that the noise aware versions perform better and are comparable. 

[compare_csp_layout.pdf](https://github.com/Qiskit/qiskit-terra/files/5304405/compare_csp_layout.pdf)

**Comment by 1ucian0** [comment]
For reference, here is the image in the PDF.
Screen Shot 2020-10-02 at 3 02 36 PM

What about runtime? Is it significantly slower than the plain CSP method?

**Comment by czachow** [comment]
Sorry for the delay.
The difference in runtime strongly depends on the number of solutions S to the layout problem. The additional time complexity can be estimated as the product of the number of gates in the circuit G and the number of solutions found by the crb solver S. I have chosen the same problem as above for an example calculation and averaged over 100 runs of the passmanager. The task is to embed a circuit of three qubits which are only interacting as a chain onto the devices.

This first plot shows the runtime for the ibmq_bogota device for the three different algorithms.
runtime_ibmq_bogota

This second plot show the runtime for the ibmq_16_melbourne device for the three different algorithms. 
runtime_ibmq_melbourne

Whereas for the bogota device there is only a small increase, the time increases dramatically for the melbourne device. In my opinion, in a typical setup the number of device and circuit qubits should be approximately equal and thus not that many different possibilities S should arise.

File: qiskit/transpiler/passes/layout/csp_layout_noise.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    def _calc_sol_gate_error(self, dag, solution):
+        """Calculate the gate error of the solution"""
+        gate_error = 0
+        for node in dag.op_nodes(include_directives=False):
+            physical_qubits = [solution[qubit.index] for qubit in node.qargs]
+            if isinstance(node.op, Gate):
+                if len(node.qargs) == 1:
+                    # exact gate not known, use average
+                    gate_error += 1/2 * (self.backend_prop.gate_error("u2", physical_qubits) +
+                                         self.backend_prop.gate_error("u3", physical_qubits))
```
Comment (1ucian0) [inline qiskit/transpiler/passes/layout/csp_layout_noise.py:108]: Recalculates the average for every unknown gate?

**Comment by 1ucian0** [comment]
After some chat with team and OP, we agree with the following:
 - Add a parameter `solution_limit` as a possible extra limit in the maximum amount of solutions that should be considered. `solution_limit=None` or `solution_limit=0` means "all the solutions". `solution_limit=1` has exactly the same effect than current `CSPLayout`.
 - Merge `CSPLayoutNoise` and `CSPLayout`.

**Comment by 1ucian0** [comment]
before this PR, we need to merge https://github.com/Qiskit/qiskit-terra/pull/6263. Therefore, `on hold`

**Comment by HuangJunye** [comment]
Just checking, what's the status of this PR? Why isn't it merged?

**Comment by czachow** [comment]
In my opinion this merge request can be closed. As a follow-on of this work we (@1ucian0, @georgios-ts and I) investigated the vf2 method for layout allocation. With merge requests #6620, #7213 and #7276 this has now become a part of qiskit.

**Comment by HuangJunye** [comment]
Luciano told me about the VF2 Layout work. Nice! Thanks for closing the PR as well.

**Pull request closed without merging**